### PR TITLE
GEM-64 | Add login endpoint

### PR DESCRIPTION
**Pull request by pwgacek**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -95,6 +95,10 @@ dependencies {
     implementation("org.springframework.boot:spring-boot-starter-security")
     implementation("org.springframework.security:spring-security-crypto")
 
+    implementation("io.jsonwebtoken:jjwt-api:0.12.5")
```
Comment (lawos98) [inline build.gradle.kts]: could you exports that deps to settings.gradle ?

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/MongoVerifiedUserRepository.kt [hunk 2]
```diff
@@ -15,4 +16,20 @@ class MongoVerifiedUserRepository(
         val query = Query().addCriteria(where("email").`is`(email))
         return mongo.exists(query, VerifiedUserEntity::class.java)
     }
+
+    override fun findByEmail(email: String): VerifiedUser? {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/MongoVerifiedUserRepository.kt:15]: IMHO its a little redundant to write function findByEmail and existByEmail if we can check if that exist using find and elvis operator

File: src/main/kotlin/pl/edu/agh/gem/security/UserDetailsServiceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import pl.edu.agh.gem.internal.persistence.VerifiedUserRepository
+
+@Service
+class UserDetailsServiceImpl(
+    private val notVerifiedUserRepository: NotVerifiedUserRepository,
+    private val verifiedUserRepository: VerifiedUserRepository,
+) : UserDetailsService {
+
+    override fun loadUserByUsername(email: String?): UserDetails {
+        if (email == null) {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/security/UserDetailsServiceImpl.kt]: if you want "email ?: throw ..."

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt [hunk 4]
```diff
@@ -153,4 +161,110 @@ class AuthControllerIT(
...
+    should("return validation exception when email is blank") {
+        // given
+        val loginRequest = createLoginRequest(email = "")
+
+        // when
+        val response = service.login(loginRequest)
+
+        // then
+        response shouldHaveHttpStatus BAD_REQUEST
+        response shouldHaveValidationError "Email can not be blank"
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt]: Could you export that to const of Test Utils?

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt [hunk 5]
```diff
@@ -153,4 +161,110 @@ class AuthControllerIT(
...
+        saveNotVerifiedUser(email = email, password = passwordEncoder.encode(password), notVerifiedUserRepository = notVerifiedUserRepository)
+        val loginRequest = createLoginRequest(email = email, password = password)
+
+        // when
+        val response = service.login(loginRequest)
+
+        // then
+        response shouldHaveHttpStatus FORBIDDEN
+        response shouldHaveErrors {
+            errors[0].code shouldBe "UserNotVerifiedException"
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt]: maybe sth UserNotVerifiedException::class.simpleName

File: src/main/kotlin/pl/edu/agh/gem/security/UserDetailsServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package pl.edu.agh.gem.security
+
+import org.springframework.security.authentication.BadCredentialsException
+import org.springframework.security.core.userdetails.UserDetails
+import org.springframework.security.core.userdetails.UserDetailsService
+import org.springframework.stereotype.Service
+import pl.edu.agh.gem.internal.persistence.NotVerifiedUserRepository
+import pl.edu.agh.gem.internal.persistence.VerifiedUserRepository
+
+@Service
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/security/UserDetailsServiceImpl.kt]: do we need to create beam if we create bean in AuthConfig UserDetailsService

File: build.gradle.kts (hunk 1 above)
Comment (lawos98) [reply build.gradle.kts]: do we want to have both?

File: src/main/kotlin/pl/edu/agh/gem/external/dto/ValidationMessage.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package pl.edu.agh.gem.external.dto
+
+class ValidationMessage private constructor() {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/dto/ValidationMessage.kt:3]: we could use object instead of class with compaion object

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/external/dto/ValidationMessage.kt:3]: will add in last pr

File: build.gradle.kts (hunk 1 above)
Comment (pwgacek) [reply build.gradle.kts]: fixed

**Review by lawos98 — APPROVED** [review]

**Pull request merged**